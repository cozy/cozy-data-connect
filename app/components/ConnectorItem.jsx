/** @jsx h */
import { h } from 'preact'
import { Link, withRouter } from 'react-router'
import { translate } from '../plugins/preact-polyglot'

const ConnectorItem = ({ title, subtitle, connected, slug, iconName, backgroundCSS = 'white', enableDefaultIcon = false, router }) => (
  <Link class='item-wrapper' to={`${router.location.pathname}/${slug}`}>
    {backgroundCSS &&
    <header class='item-header' style={{background: backgroundCSS}}>
      {iconName &&
        <svg class='item-icon'>
          <use xlinkHref={icon(iconName, enableDefaultIcon)} />
        </svg>
      }
    </header>
    }
    {!backgroundCSS &&
    <header class='item-header'>
      {iconName &&
        <svg class='item-icon'>
          <use xlinkHref={icon(iconName, enableDefaultIcon)} />
        </svg>
      }
    </header>
    }
    <p class='item-title'>{title}</p>
    {subtitle && <p class='item-subtitle'>{subtitle}</p>}
    {connected &&
      <svg class='item-connected'>
        <use xlinkHref={require('../assets/sprites/icon-check.svg')} />
      </svg>
    }
  </Link>
)

// Fallback to get the item icon and avoid error if not found
// with a possible default icon
const icon = (iconName, enableDefaultIcon) => {
  let icon = ''
  try {
    icon = require(`../assets/icons/${iconName}.svg`)
  } catch (e) {
    if (enableDefaultIcon) {
      icon = require('../assets/icons/default_myaccount.svg')
    }
  }
  return icon
}

export default translate()(withRouter(ConnectorItem))
