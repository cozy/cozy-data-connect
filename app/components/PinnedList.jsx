/** @jsx h */
import { h } from 'preact'
import { translate } from '../plugins/preact-polyglot'
import ConnectorList from './ConnectorList'

const PinnedList = ({ t, connectors, children }) => (
  <div class='content'>
    <h1>{t('my_accounts pinned title')}</h1>
    <ConnectorList connectors={connectors} />
    {children}
  </div>
)

export default translate()(PinnedList)
