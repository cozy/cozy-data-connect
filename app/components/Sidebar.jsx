/** @jsx h */
import { h } from 'preact'
import { Link, withRouter } from 'react-router'
import { translate } from '../plugins/preact-polyglot'

const Sidebar = ({ t, categories, router }) => {
  let isCategoryView = router.location.pathname.match(/^\/category/) !== null
  const i18nCategories =
    categories.filter(c => c !== 'others')
      .map(c => ({ slug: c, label: t(`${c} category`) }))
      .sort((a, b) => a.label.localeCompare(b.label))

  return (
    <aside className='coz-sidebar'>
      <ul role='navigation' className='coz-nav'>
        <li className='coz-nav-item'>
          <Link to='/discovery' className='coz-nav-link con-cat-discovery' activeClassName='active'>
            {t('my_accounts discovery title')}
          </Link>
        </li>
        <li className='coz-nav-item'>
          <Link to='/category/all'
            className={isCategoryView ? 'coz-nav-link con-cat-categories active' : 'coz-nav-link con-cat-categories'}>
            {t('my_accounts category title')}
          </Link>
          {isCategoryView &&
            <ul className='con-nav-submenu'>
              <li>
                <Link to='/category/all' className='con-nav-submenu-link' activeClassName='active'>
                  {t('all category')}
                </Link>
              </li>
              {i18nCategories.map(category => (
                <li>
                  <Link to={`/category/${category.slug}`} className='con-nav-submenu-link' activeClassName='active'>
                    {category.label}
                  </Link>
                </li>
              ))}
              <li>
                <Link to='/category/others' className='con-nav-submenu-link' activeClassName='active'>
                  {t('others category')}
                </Link>
              </li>
            </ul>
          }
        </li>
        <li className='coz-nav-item'>
          <Link to='/connected' className='coz-nav-link con-cat-pinned' activeClassName='active'>
            {t('my_accounts connected title')}
          </Link>
        </li>
      </ul>
    </aside>
  )
}
export default translate()(withRouter(Sidebar))
