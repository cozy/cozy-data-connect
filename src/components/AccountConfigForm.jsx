import React from 'react'

import Field, { DropdownField } from './Field'
import AccountLoginForm from './AccountLoginForm'

// Ensure that path has a / as first character
const sanitizePath = path => path.charAt(0) === '/' ? path : `/${path}`

const folderLink = (path, folders) => {
  path = sanitizePath(path)
  let folder = folders.find(f => f.path + '/' + f.name === path)
  if (!folder) return
  return `/apps/files/#folders/${folder.id}`
}

const AccountConfigForm = ({ t, customView, fields, dirty, error, submit, submitting }) => (
  <div className={'account-form' + (error ? ' error' : '')}>
    {fields.folderPath &&
      <div>
        <h3>{t('location')}</h3>
        <p>{t('location desc')}</p>
        <DropdownField label={false} {...fields.folderPath} />
        <a href={folderLink(fields.folderPath.value, fields.folderPath.folders)} target='_blank'>
          {t('location button')}
        </a>
      </div>
    }
    {fields.calendar &&
      <div>
        <h3>{t('calendar')}</h3>
        <p>{t('calendar desc')}</p>
        <Field label={false} {...fields.calendar} />
      </div>
    }
    <h3>{t('frequency')}</h3>
    <p>{t('frequency desc')}</p>
    <Field label={false} {...fields.frequency}>
      <select className='account-field-dropdown'>
        <option value='none'>{t('none')}</option>
        <option value='hour'>{t('every hour')}</option>
        <option value='day'>{t('every day')}</option>
        <option value='week'>{t('every week')}</option>
        <option value='month'>{t('each month')}</option>
      </select>
    </Field>
    <h3>{t('account')}</h3>
    <AccountLoginForm
      t={t}
      customView={customView}
      fields={fields}
    />
    {error === 'bad credentials' &&
      <p className='errors'>{t('account config bad credentials')}</p>
    }
  </div>
)

export default AccountConfigForm
