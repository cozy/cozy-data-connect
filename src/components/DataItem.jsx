import React from 'react'
import { translate } from '../plugins/i18n'

const DataItem = ({ t, dataType, hex }) => (
  <li className='account-data'>
    <svg className='account-dataIcon' style={{color: hex}}>
      <use xlinkHref={require('../assets/sprites/icon-' + dataType + '.svg')} />
    </svg>
    {t('dataType ' + dataType)}
  </li>
)

export default translate()(DataItem)
