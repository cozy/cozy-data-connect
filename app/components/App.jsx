/** @jsx h */
import { h } from 'preact'

import styles from '../styles/app.styl'
import classNames from 'classnames'

import Sidebar from './Sidebar'
import Notifier from './Notifier'

const App = ({ categories, children }) => (
  <div className={classNames(styles['con-wrapper'], styles['coz-sticky'])}>
    <Sidebar categories={categories} />
    <main className={styles['con-content']}>
      {children}
    </main>
    <Notifier />
  </div>
)

export default App
