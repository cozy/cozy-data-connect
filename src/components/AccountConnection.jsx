import React, { Component } from 'react'
import statefulForm from '../lib/statefulForm'

import AccountLoginForm from './AccountLoginForm'
import DataItem from './DataItem'
import ReactMarkdown from 'react-markdown'

class AccountConnection extends Component {
  constructor (props, context) {
    super(props, context)

    this.state = {
      submitting: false
    }

    this.submit = () => {
      this.setState({
        submitting: true
      })

      this.props.submit()
        .catch(error => this.setState({submitting: false, error: error.message}))
    }
  }

  render () {
    const { t, connector, fields, error } = this.props
    let { dirty } = this.props
    const { name, description } = connector
    const { submitting } = this.state

    let button = <button disabled={!dirty} aria-busy={submitting ? 'true' : 'false'}
      onClick={() => this.submit()}>
      {t('account config button')}
    </button>

    if (connector.oauth) {
      // TODO find properly the root cozy url (without data-account)
      button = <a target='_blank' href='http://cozy.tools:8080/accounts/orange/start'>Submit</a>
    }

    // If there is no field displayed, the form is dirty by default.
    dirty = dirty || Object.values(fields).every(field => field.type === 'hidden' || field.advanced)
    return (
      <div className='account-connection'>
        <div className='account-description'>
          <h3>{t('title description')}</h3>
          <p>
            <ReactMarkdown
              source={
                t(description)
              }
              renderers={{Link: props => <a href={props.href} target='_blank'>{props.children}</a>}}
            />
          </p>
          <h3>{t('dataType title')}</h3>
          <ul className='account-datas'>
            {connector.dataType && connector.dataType.map(data =>
              <DataItem
                dataType={data}
                hex={connector.color.hex}
              />
            )}
          </ul>
          <p>{` ${connector.name} ${t('dataType disclaimer')} `}</p>
        </div>
        <div className='account-login'>
          <h3>{t('account config title', {name: name})}</h3>
          <div className={'account-form' + (error ? ' error' : '')}>
            <AccountLoginForm
              t={t}
              fields={fields}
            />
            <div className='account-form-controls'>
              {button}
              {error === 'bad credentials' &&
                <p className='errors'>{t('account config bad credentials')}</p>
              }
            </div>
          </div>
        </div>
      </div>
    )
  }
}

export default statefulForm()(AccountConnection)
