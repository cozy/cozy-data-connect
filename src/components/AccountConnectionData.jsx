import styles from '../styles/accountConnectionData'

import React from 'react'
import { translate } from 'cozy-ui/react/I18n'
import ClassNames from 'classnames'
import ReactMarkdownWrapper from '../components/ReactMarkdownWrapper'
import DataItem from '../components/DataItem'

const AccountConnectionData = ({ t, connector }) => {
  const { hasDescriptions } = connector
  const hasDataTypes = !!(connector.dataType && connector.dataType.length)

  return (
    <div className={ClassNames(styles['col-account-connection-data'])}>
      {hasDescriptions &&
        hasDescriptions.service && (
          <div>
            <h4>{t('account.config.data.service.description')}</h4>
            <p>
              <ReactMarkdownWrapper
                source={t(`connector.${connector.slug}.description.service`)}
              />
            </p>
          </div>
        )}
      <h4>{t('account.config.data.title')}</h4>
      {hasDataTypes && (
        <ul className={styles['col-account-connection-data-access']}>
          {connector.dataType.map(data => (
            <DataItem key={data} dataType={data} />
          ))}
        </ul>
      )}
      {!hasDataTypes && <p>{t('dataType.none', { name: connector.name })}</p>}
    </div>
  )
}

export default translate()(AccountConnectionData)
