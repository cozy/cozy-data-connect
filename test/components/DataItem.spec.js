'use strict'

/* eslint-env jest */

import React from 'react'
import { configure, shallow } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'

import { tMock } from '../jestLib/I18n'
import { DataItem } from '../../src/components/DataItem'

import datatypesToTest from '../../src/config/datatypes'

const hexColorMock = '#297EF2'

configure({ adapter: new Adapter() })

describe('DataItem component', () => {
  beforeEach(() => {
    jest.resetModules()
  })

  for (const dataType of datatypesToTest) {
    it(`should display correctly the ${dataType} data type`, () => {
      const component = shallow(
        <DataItem t={tMock} dataType={dataType} hex={hexColorMock} />
      ).getElement()
      expect(component).toMatchSnapshot()
    })
  }
})
