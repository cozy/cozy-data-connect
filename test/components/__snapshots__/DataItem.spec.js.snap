// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataItem component should display correctly the activity data type 1`] = `
<li
  className="col-data-item"
>
  <Icon
    className="col-data-item-icon"
    height="32px"
    icon="test-file-stub"
    spin={false}
    width="32px"
  />
  Your activities
</li>
`;

exports[`DataItem component should display correctly the appointment data type 1`] = `
<li
  className="col-data-item"
>
  <Icon
    className="col-data-item-icon"
    height="32px"
    icon="test-file-stub"
    spin={false}
    width="32px"
  />
  Your appointments
</li>
`;

exports[`DataItem component should display correctly the bankAccounts data type 1`] = `
<li
  className="col-data-item"
>
  <Icon
    className="col-data-item-icon"
    height="32px"
    icon="test-file-stub"
    spin={false}
    width="32px"
  />
  Your bank accounts
</li>
`;

exports[`DataItem component should display correctly the bankTransactions data type 1`] = `
<li
  className="col-data-item"
>
  <Icon
    className="col-data-item-icon"
    height="32px"
    icon="test-file-stub"
    spin={false}
    width="32px"
  />
  Your bank transactions
</li>
`;

exports[`DataItem component should display correctly the bill data type 1`] = `
<li
  className="col-data-item"
>
  <Icon
    className="col-data-item-icon"
    height="32px"
    icon="test-file-stub"
    spin={false}
    width="32px"
  />
  Your bills
</li>
`;

exports[`DataItem component should display correctly the bloodPressure data type 1`] = `
<li
  className="col-data-item"
>
  <Icon
    className="col-data-item-icon"
    height="32px"
    icon="test-file-stub"
    spin={false}
    width="32px"
  />
  Your blood pressure
</li>
`;

exports[`DataItem component should display correctly the calendar data type 1`] = `
<li
  className="col-data-item"
>
  <Icon
    className="col-data-item-icon"
    height="32px"
    icon="test-file-stub"
    spin={false}
    width="32px"
  />
  Your calendars
</li>
`;

exports[`DataItem component should display correctly the certificate data type 1`] = `
<li
  className="col-data-item"
>
  <Icon
    className="col-data-item-icon"
    height="32px"
    icon="test-file-stub"
    spin={false}
    width="32px"
  />
  Your certificates
</li>
`;

exports[`DataItem component should display correctly the commit data type 1`] = `
<li
  className="col-data-item"
>
  <Icon
    className="col-data-item-icon"
    height="32px"
    icon="test-file-stub"
    spin={false}
    width="32px"
  />
  Your commits
</li>
`;

exports[`DataItem component should display correctly the consumption data type 1`] = `
<li
  className="col-data-item"
>
  <Icon
    className="col-data-item-icon"
    height="32px"
    icon="test-file-stub"
    spin={false}
    width="32px"
  />
  Your consumption
</li>
`;

exports[`DataItem component should display correctly the contact data type 1`] = `
<li
  className="col-data-item"
>
  <Icon
    className="col-data-item-icon"
    height="32px"
    icon="test-file-stub"
    spin={false}
    width="32px"
  />
  Your contacts
</li>
`;

exports[`DataItem component should display correctly the contract data type 1`] = `
<li
  className="col-data-item"
>
  <Icon
    className="col-data-item-icon"
    height="32px"
    icon="test-file-stub"
    spin={false}
    width="32px"
  />
  Your contracts
</li>
`;

exports[`DataItem component should display correctly the courseMaterial data type 1`] = `
<li
  className="col-data-item"
>
  <Icon
    className="col-data-item-icon"
    height="32px"
    icon="test-file-stub"
    spin={false}
    width="32px"
  />
  Your course materials
</li>
`;

exports[`DataItem component should display correctly the event data type 1`] = `
<li
  className="col-data-item"
>
  <Icon
    className="col-data-item-icon"
    height="32px"
    icon="test-file-stub"
    spin={false}
    width="32px"
  />
  Your events
</li>
`;

exports[`DataItem component should display correctly the family data type 1`] = `
<li
  className="col-data-item"
>
  <Icon
    className="col-data-item-icon"
    height="32px"
    icon="test-file-stub"
    spin={false}
    width="32px"
  />
  Your family informations
</li>
`;

exports[`DataItem component should display correctly the geopoint data type 1`] = `
<li
  className="col-data-item"
>
  <Icon
    className="col-data-item-icon"
    height="32px"
    icon="test-file-stub"
    spin={false}
    width="32px"
  />
  Your geolocations
</li>
`;

exports[`DataItem component should display correctly the heartbeat data type 1`] = `
<li
  className="col-data-item"
>
  <Icon
    className="col-data-item-icon"
    height="32px"
    icon="test-file-stub"
    spin={false}
    width="32px"
  />
  Your heartbeats
</li>
`;

exports[`DataItem component should display correctly the home data type 1`] = `
<li
  className="col-data-item"
>
  <Icon
    className="col-data-item-icon"
    height="32px"
    icon="test-file-stub"
    spin={false}
    width="32px"
  />
  Your home informations
</li>
`;

exports[`DataItem component should display correctly the phonecommunicationlog data type 1`] = `
<li
  className="col-data-item"
>
  <Icon
    className="col-data-item-icon"
    height="32px"
    icon="test-file-stub"
    spin={false}
    width="32px"
  />
  Your phone communications log
</li>
`;

exports[`DataItem component should display correctly the podcast data type 1`] = `
<li
  className="col-data-item"
>
  <Icon
    className="col-data-item-icon"
    height="32px"
    icon="test-file-stub"
    spin={false}
    width="32px"
  />
  Your podcasts
</li>
`;

exports[`DataItem component should display correctly the profile data type 1`] = `
<li
  className="col-data-item"
>
  <Icon
    className="col-data-item-icon"
    height="32px"
    icon="test-file-stub"
    spin={false}
    width="32px"
  />
  Your client profile
</li>
`;

exports[`DataItem component should display correctly the refund data type 1`] = `
<li
  className="col-data-item"
>
  <Icon
    className="col-data-item-icon"
    height="32px"
    icon="test-file-stub"
    spin={false}
    width="32px"
  />
  Your refunds
</li>
`;

exports[`DataItem component should display correctly the sinister data type 1`] = `
<li
  className="col-data-item"
>
  <Icon
    className="col-data-item-icon"
    height="32px"
    icon="test-file-stub"
    spin={false}
    width="32px"
  />
  Your sinisters
</li>
`;

exports[`DataItem component should display correctly the sleepTime data type 1`] = `
<li
  className="col-data-item"
>
  <Icon
    className="col-data-item-icon"
    height="32px"
    icon="test-file-stub"
    spin={false}
    width="32px"
  />
  Your sleep time
</li>
`;

exports[`DataItem component should display correctly the stepsNumber data type 1`] = `
<li
  className="col-data-item"
>
  <Icon
    className="col-data-item-icon"
    height="32px"
    icon="test-file-stub"
    spin={false}
    width="32px"
  />
  Your number of steps
</li>
`;

exports[`DataItem component should display correctly the temperature data type 1`] = `
<li
  className="col-data-item"
>
  <Icon
    className="col-data-item-icon"
    height="32px"
    icon="test-file-stub"
    spin={false}
    width="32px"
  />
  Your temperature data
</li>
`;

exports[`DataItem component should display correctly the travelDate data type 1`] = `
<li
  className="col-data-item"
>
  <Icon
    className="col-data-item-icon"
    height="32px"
    icon="test-file-stub"
    spin={false}
    width="32px"
  />
  Your travel dates
</li>
`;

exports[`DataItem component should display correctly the tweet data type 1`] = `
<li
  className="col-data-item"
>
  <Icon
    className="col-data-item-icon"
    height="32px"
    icon="test-file-stub"
    spin={false}
    width="32px"
  />
  Your tweets
</li>
`;

exports[`DataItem component should display correctly the videostream data type 1`] = `
<li
  className="col-data-item"
>
  <Icon
    className="col-data-item-icon"
    height="32px"
    icon="test-file-stub"
    spin={false}
    width="32px"
  />
  Your videos
</li>
`;

exports[`DataItem component should display correctly the weight data type 1`] = `
<li
  className="col-data-item"
>
  <Icon
    className="col-data-item-icon"
    height="32px"
    icon="test-file-stub"
    spin={false}
    width="32px"
  />
  Your weights
</li>
`;
